crates/models/llama/src/lib.rs

```diff
@@ -52,39 +63,54 @@ impl KnownModel for Llama {
 
         // model-global weights
         let wte = tl.load("tok_embeddings.weight")?;
-        let norm = tl.load("norm.weight")?;
-        let output = tl.load("output.weight")?;
+
+        let backend = if params.should_offload(0) {
+            Backend::Gpu
+        } else {
+            Backend::Cpu
+        };
+
+        let norm = tl.offload("norm.weight", backend)?;
+
+        let output = tl.offload("output.weight", backend)?;
 
         let mut layers = Vec::new();
+
         for i in 0..hyperparameters.n_layer {
+            let backend = if params.should_offload(i) {
```

Comment:
Okay, maybe `params.backend(layer_index: usize)`.

As an aside - because `should_offload(0)` is used for both `norm/output`/the rest of the layer, it means that layer 0 will be larger than the others. Not sure if it matters too much.

Reply:
`params.backend(layer_index: usize)` Sounds like a good idea. `llama.cpp` handles the `norm/output` by checking if the number of layers that should be offloaded is >= the number of layers in the model. If it is they will be offloaded. I decided to offload them if at least one layer is offloaded, maybe we should implement it similar to `llama.cpp` to avoid unnecessary copying from ram->gpu->ram->gpu 🤔